### PR TITLE
[AKS] Add "--skip-role-assignment" flag and support creation on same role assignment

### PR DESCRIPTION
**Pull request by zqingqing1**
This PR mainly has following changes:
1) add '--skip-role-assignment' which allows customer to skip role assignment for customVnet when creating AKS cluster.
2)If role assignment for customVnet  already exists, new AKS creation will not fail, and continue provisioning.
3) remove non-used fields in AgentPoolProfile

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py [hunk 1]
```diff
@@ -152,6 +152,10 @@
           short-summary: Service principal used for authentication to Azure APIs.
           long-summary:  If not specified, a new service principal is created and cached at
                          {sp_cache} to be used by subsequent `az aks` commands.
+        - name: --skip-role-assignemnt
```
Comment (weinong) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py]: should this be more specific?

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 2]
```diff
@@ -1397,6 +1412,12 @@ def aks_create(cmd, client, resource_group_name, name, ssh_key_value,  # pylint:
         secret=principal_obj.get("client_secret"),
         key_vault_secret_ref=None)
 
+    if vnet_subnet_id and not skip_role_assignment and not role_assignment_exist(cmd.cli_ctx, vnet_subnet_id):
+        scope = vnet_subnet_id
+        if not _add_role_assignment(cmd.cli_ctx, 'Contributor', service_principal, scope=scope):
```
Comment (weinong) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: s/Contributor/Network Contributor/

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 3]
```diff
@@ -1397,6 +1412,12 @@ def aks_create(cmd, client, resource_group_name, name, ssh_key_value,  # pylint:
...
+            logger.warning('Could not create a role assignment. '
+                           'Are you an Owner on this project?')
```
Comment (weinong) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: s/project/subscription/

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 4]
```diff
@@ -1287,6 +1290,18 @@ def _update_dict(dict1, dict2):
     return cp
 
 
+def role_assignment_exist(cli_ctx, scope):
```
Comment (weinong) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: do you not need to check the role?

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py [hunk 5]
```diff
@@ -173,6 +173,7 @@ def load_arguments(self, _):
         c.argument('service_cidr')
         c.argument('vnet_subnet_id')
         c.argument('workspace_resource_id')
+        c.argument('skip_role_assignment', required=False, action='store_true')
```
Comment (weinong) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py]: what does `action='store_true'` mean?

File: src/command_modules/azure-cli-acs/HISTORY.rst [hunk 6]
```diff
@@ -3,6 +3,11 @@
 Release History
 ===============
 
+2.2.3
++++++
+* Add new option "skip role assignment" for customVnet when creating
+* Skip role assignment when it exists instead of return error
```
Comment (weinong) [inline src/command_modules/azure-cli-acs/HISTORY.rst]: you also create role assignment to the subnet when it's provided.

**Comment by weinong** [comment]
do we still create subscription scope role assignment when SP is not provided? if so, can you open a bug?

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py (hunk 5 above)
Comment (zqingqing1) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py]: when specified, the value for the flag is true.
```                    details['options']['action'] = 'store_true'
                    self._help(name, "True if flag present.")
```

**Comment by zqingqing1** [comment]
We don't create role assignment when SP not provided when use `az aks create`

File: sp.txt [hunk 7]
```diff
@@ -0,0 +1,22 @@
+a45cf3fb-0d96-40eb-994e-463af028d244
```
Comment (weinong) [inline sp.txt]: ??

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py [hunk 8]
```diff
@@ -152,6 +152,10 @@
           short-summary: Service principal used for authentication to Azure APIs.
           long-summary:  If not specified, a new service principal is created and cached at
                          {sp_cache} to be used by subsequent `az aks` commands.
+        - name: --skip-subnet-role-assignment
+          type: bool
+          short-summary: Skip role assignment for custom vnet (advanced networking).
```
Comment (weinong) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py]: s/for custom vnet/to subnet/

File: src/command_modules/azure-cli-acs/HISTORY.rst [hunk 9]
```diff
@@ -3,6 +3,12 @@
 Release History
 ===============
 
+2.2.3
++++++
+* Create role assignment to the subnet when it's provided.
+* Add new option "skip role assignment" for customVnet when it's provided
+* Skip role assignment when it exists instead of return error
```
Comment (weinong) [inline src/command_modules/azure-cli-acs/HISTORY.rst]: Skip role assignment to subnet when assignment has already existed

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 10]
```diff
@@ -1397,6 +1414,13 @@ def aks_create(cmd, client, resource_group_name, name, ssh_key_value,  # pylint:
         secret=principal_obj.get("client_secret"),
         key_vault_secret_ref=None)
 
+    if (vnet_subnet_id and not skip_subnet_role_assignment and
+            not subnet_role_assignment_exists(cmd.cli_ctx, vnet_subnet_id)):
+        scope = vnet_subnet_id
+        if not _add_role_assignment(cmd.cli_ctx, 'Network Contributor', service_principal, scope=scope):
+            logger.warning('Could not create a role assignment for custom vnet. '
```
Comment (weinong) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: s/for custom vnet/to the subnet/

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py [hunk 11]
```diff
@@ -202,7 +202,77 @@ def test_aks_bad_inputs(self, resource_group, resource_group_location, sp_name,
...
+    def test_aks_create_default_service_with_skip_role_assignment(self, resource_group, resource_group_location, sp_name, sp_password):
+        aks_name = self.create_random_name('cliakstest', 16)
+        self.kwargs.update({
+            'resource_group': resource_group,
+            'name': aks_name,
+            'location': resource_group_location,
+            'service_principal': sp_name,
+            'client_secret': sp_password,
+            'vnet_subnet_id': self.generate_vnet_subnet_id(resource_group),
+            'resource_type': 'Microsoft.ContainerService/ManagedClusters'
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py]: You don't use the `resource_type` kwarg, you can remove it.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py [hunk 12]
```diff
@@ -202,7 +202,77 @@ def test_aks_bad_inputs(self, resource_group, resource_group_location, sp_name,
...
+    def test_aks_create_default_service_without_skip_role_assignment(self, resource_group, resource_group_location, sp_name, sp_password):
+        aks_name = self.create_random_name('cliakstest', 16)
+        self.kwargs.update({
+            'resource_group': resource_group,
+            'name': aks_name,
+            'location': resource_group_location,
+            'service_principal': sp_name,
+            'client_secret': sp_password,
+            'vnet_subnet_id': self.generate_vnet_subnet_id(resource_group),
+            'resource_type': 'Microsoft.ContainerService/ManagedClusters'
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py]: You don't use this.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 13]
```diff
@@ -510,7 +510,10 @@ def _add_role_assignment(cli_ctx, role, service_principal, delay=2):
         hook.add(message='Waiting for AAD role to propagate', value=0.1 * x, total_val=1.0)
         try:
             # TODO: break this out into a shared utility library
-            create_role_assignment(cli_ctx, role, service_principal)
+            if not scope:
+                create_role_assignment(cli_ctx, role, service_principal)
+            else:
+                create_role_assignment(cli_ctx, role, service_principal, scope=scope)
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Combine these into `create_role_assignment(cli_ctx, role, service_principal, scope=scope)`.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py [hunk 14]
```diff
@@ -152,6 +152,10 @@
           short-summary: Service principal used for authentication to Azure APIs.
           long-summary:  If not specified, a new service principal is created and cached at
                          {sp_cache} to be used by subsequent `az aks` commands.
+        - name: --skip-subnet-role-assignment
+          type: bool
+          short-summary: Skip role assignment for subnet (advanced networking).
+          long-summary:  If it is true, please make sure your service principal has the access to your subnet.
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py]: If specifying this, please make...

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py [hunk 15]
```diff
@@ -173,6 +173,7 @@ def load_arguments(self, _):
         c.argument('service_cidr')
         c.argument('vnet_subnet_id')
         c.argument('workspace_resource_id')
+        c.argument('skip-subnet-role-assignment', required=False, action='store_true')
```
Comment (williexu) [inline src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py]: remove the `required=False`, this will be inferred from the function signature.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 16]
```diff
@@ -1287,6 +1290,20 @@ def _update_dict(dict1, dict2):
...
 
 
+def subnet_role_assignment_exists(cli_ctx, scope):
+    exist = False
+    network_contributor_role_id = "4d97b98b-1d4f-4787-a291-c67834d212e7"
+
+    factory = get_auth_management_client(cli_ctx, scope)
+    assignments_client = factory.role_assignments
+
+    for i in list(assignments_client.list_for_scope(scope=scope, filter='atScope()')):
```
[the remark this answers is not in the document]
Comment (zqingqing1) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: yes, we need to check item in the list to find the one we need.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py [hunk 17]
```diff
@@ -1287,6 +1290,20 @@ def _update_dict(dict1, dict2):
...
+        if i.scope == scope and i.role_definition_id.endswith(network_contributor_role_id):
+            exist = True
+            break
```
[the remark this answers is not in the document]
Comment (zqingqing1) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Make sense. I was using 'exist' when I need to do some printing when troubleshooting

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py (hunk 11 above)
Comment (zqingqing1) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py]: why not need this? is this will be the default one?

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py (hunk 15 above)
Comment (zqingqing1) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_params.py]: Ok, don't know this. will make a change

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py (hunk 14 above)
Comment (williexu) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py]: Oops sorry I meant the long-summary text itself. Full sentence:
`If specifying this, please make sure your service principal has the access to your subnet.`

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py (hunk 11 above)
Comment (zqingqing1) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py]: never mind, you are right.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (williexu) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: I meant does `list_for_scope` already return an iterable? If so, you dont need `list()`

[the remark this answers is not in the document]
Comment (zqingqing1) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: I don't think so. If you searched the whole repo, we also find that `azure-cli-role` using the same syntax. BTW, I didn't find the 'list_for_scope()' function signature, which gives me hard time. If you can find that, please do let me know.

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py (hunk 14 above)
Comment (zqingqing1) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py]: ok, got you

File: src/command_modules/azure-cli-acs/HISTORY.rst [hunk 18]
```diff
@@ -3,6 +3,12 @@
 Release History
 ===============
 
+2.2.3
++++++
+* Create role assignment to the subnet when it's provided.
+* Add new option "skip role assignment" for subnet when it's provided
+* Skip role assignment to subnet when assignment has already existed
```
Comment (mboersma) [inline src/command_modules/azure-cli-acs/HISTORY.rst]: Nitpick: I would reword the last bullet point to say
`* Skip role assignment to subnet when assignment already exists`

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py (hunk 13 above)
Comment (mboersma) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: It doesn't look like the `if not scope:` test is necessary, I agree. Is the default value for `scope` not `None` in that function?

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mboersma) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: Here is [the definition of `list_for_scope`](https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt-authorization/azure/mgmt/authorization/v2015_07_01/operations/role_assignments_operations.py#L674). I think it should return an iterable--could you try without `list()`?

[the remark this answers is not in the document]
Comment (williexu) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: That's because it is coming from a sdk dependency: https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt-authorization/azure/mgmt/authorization/v2018_01_01_preview/operations/role_assignments_operations.py#L661-L677

File: src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py (hunk 13 above)
Comment (williexu) [reply src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py]: That should work; [create_role_assignment](https://github.com/Azure/azure-cli/blob/dev/src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py#L1206) has the following signature:
`def create_role_assignment(cli_ctx, role, assignee, resource_group_name=None, scope=None):`

If scope is None, it will just set that.

**Comment by williexu** [comment]
@zqingqing1 one last thing, can you rebase from dev to get updates?
